pathd/path_ted.c

```diff
@@ -0,0 +1,693 @@
+			ted_state_g.import = IMPORT_UNKNOWN;
+		}
+	}
+	return status;
+}
+
+/**
+ * Unset all needed to receive igp data.
+ *
+ * @return		true if ok
+ *
+ */
+uint32_t path_ted_stop_importing_igp(void)
+{
+	uint32_t status = 0;
+
+	if (ted_state_g.import != IMPORT_UNKNOWN) {
+		if (ls_unregister(zclient, false /*client*/) != 0) {
+			zlog_err(
+				"%s: PATHD-TED: Unable to unregister Link State",
+				__func__);
+			status = 1;
+		} else {
+			ted_state_g.import = IMPORT_UNKNOWN;
+			zlog_debug("%s: PATHD-TED: Importing igp data OFF",
+				   __func__);
+			path_ted_timer_sync_cancel();
+		}
+	}
+	return status;
+}
+/**
+ * Check for ted status
+ *
+ * @return		true if ok
+ *
+ */
+bool path_ted_is_initialized(void)
+{
+	if (ted_state_g.ted == NULL) {
+		zlog_warn("PATHD TED ls_ted not initialized");
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Internal util functions
+ */
+
+/**
+ * Creates an empty ted
+ *
+ * @param void
+ *
+ * @return		Ptr to ted or NULL
+ */
+struct ls_ted *path_ted_create_ted()
+{
+	struct ls_ted *ted = ls_ted_new(TED_KEY, TED_NAME, TED_ASN);
+
+	if (ted == NULL) {
+		zlog_err("%s Unable to initialize TED Key [%d] ASN [%d] Name [%s]",
+			 __func__, TED_KEY, TED_ASN, TED_NAME);
+	} else {
+		zlog_info("%s Initialize TED Key [%d] ASN [%d] Name [%s]",
+			 __func__, TED_KEY, TED_ASN, TED_NAME);
+	}
+
+	return ted;
+}
+
+uint32_t path_ted_rcvd_message(struct ls_message *msg)
+{
+	if (!path_ted_is_initialized())
+		return MPLS_LABEL_NONE;
+
+	if (msg == NULL) {
+		zlog_err("%s: [rcv ted] TED received NULL message ", __func__);
+		return 1;
+	}
+
+	if (path_ted_get_current_igp(msg->data.node->adv.origin))
+		return 1;
+
+	switch (msg->type) {
+	case LS_MSG_TYPE_NODE:
+		ls_msg2vertex(ted_state_g.ted, msg, true /*hard delete*/);
+		break;
+
+	case LS_MSG_TYPE_ATTRIBUTES:
+		ls_msg2edge(ted_state_g.ted, msg, true /*ĥard delete*/);
+		break;
+
+	case LS_MSG_TYPE_PREFIX:
+		ls_msg2subnet(ted_state_g.ted, msg, true /*hard delete*/);
+		break;
+
+	default:
+		zlog_debug(
+			"%s: [rcv ted] TED received unknown message type [%d]",
+			__func__, msg->type);
+		break;
+	}
+	return 0;
+}
+
+uint32_t path_ted_query_type_f(struct ipaddr *local, struct ipaddr *remote)
+{
+	uint32_t sid = MPLS_LABEL_NONE;
+	struct ls_edge *edge;
+
+	if (!path_ted_is_initialized())
+		return MPLS_LABEL_NONE;
+
+	switch (local->ipa_type) {
+	case IPADDR_V4:
+		/* We have local and remote ip */
+		/* so check all attributes in ted */
+		frr_each (edges, &ted_state_g.ted->edges, edge) {
+			if (edge->attributes->standard.local.s_addr
+				    == local->ip._v4_addr.s_addr
+			    && edge->attributes->standard.remote.s_addr
+				       == remote->ip._v4_addr.s_addr
+			    && CHECK_FLAG(edge->attributes->flags,
+					  LS_ATTR_ADJ_SID)) {
+				sid = edge->attributes->adj_sid[0]
+					      .sid; /* from primary */
+				break;
+			}
+		}
+		break;
+	case IPADDR_V6:
+		frr_each (edges, &ted_state_g.ted->edges, edge) {
```

Comment:
TED uses RB_TREE. Don't need to loop other all edges. Just query the edge from local IP address with _find function. edge key is base on local ip address. Please have a look to lib/link_state.c as example